src/lxc/network.c

```diff
@@ -376,6 +376,150 @@ static int instantiate_macvlan(struct lxc_handler *handler, struct lxc_netdev *n
 	return -1;
 }
 
+static int lxc_ipvlan_create(const char *master, const char *name, int mode, int isolation)
+{
+	int err, index, len;
+	struct ifinfomsg *ifi;
+	struct nl_handler nlh;
+	struct rtattr *nest, *nest2;
+	struct nlmsg *answer = NULL, *nlmsg = NULL;
+
+	err = netlink_open(&nlh, NETLINK_ROUTE);
+	if (err)
+		return err;
+
+	err = -EINVAL;
+	len = strlen(master);
+	if (len == 1 || len >= IFNAMSIZ)
+		goto out;
+
+	len = strlen(name);
+	if (len == 1 || len >= IFNAMSIZ)
+		goto out;
+
+	err = -ENOMEM;
+	nlmsg = nlmsg_alloc(NLMSG_GOOD_SIZE);
+	if (!nlmsg)
+		goto out;
+
+	answer = nlmsg_alloc_reserve(NLMSG_GOOD_SIZE);
+	if (!answer)
+		goto out;
+
+	err = -EINVAL;
+	index = if_nametoindex(master);
+	if (!index)
+		goto out;
+
+	nlmsg->nlmsghdr->nlmsg_flags =
+	    NLM_F_REQUEST | NLM_F_CREATE | NLM_F_EXCL | NLM_F_ACK;
+	nlmsg->nlmsghdr->nlmsg_type = RTM_NEWLINK;
+
+	ifi = nlmsg_reserve(nlmsg, sizeof(struct ifinfomsg));
+	if (!ifi) {
+		err = -ENOMEM;
+		goto out;
+	}
+	ifi->ifi_family = AF_UNSPEC;
+
+	nest = nla_begin_nested(nlmsg, IFLA_LINKINFO);
```

Comment:
Is `err = -EINVAL` still valid here?

Reply:
Actually same questions for all codepaths after that.

Reply:
@brauner some of the conventions here have me a bit confused. Knowing when a function should set `errno` directly and return -1, vs returning the err value from another function call (and whether it should be returned converted to a minus value) is still alluding me if I'm honest.

There seems to be various approaches being used (at least in network.c).

However I see what you are getting at, so I've had a go at re-arranging this function to ensure that `err` (and by extension `errno` when used with `minus_one_set_errno()`) will be set to a more appropriate value.

I've moved some of the basic input validation upwards before netlink resource allocation so that I can use minus_one_set_errno(EINVAL) directly without having to worry about freeing resources.

Then looking at the source of `netlink_open`, that returns either `-errno` or a specific error value with a minus.  

https://github.com/lxc/lxc/blob/master/src/lxc/nl.c#L337

So I've undone the minus when passing that returned err value into `minus_one_set_errno()`. That seems to be the done thing.

After that I've left the err being set to -ENOMEM for the 3 memory allocation steps as that seems appropriate.

And then as the netlink message is constructed I've switched err to be -EPROTO, I'd be interested to know whether you feel this is appropriate.

Finally, at the bottom in the `out:` block, I needed to ensure that the resources are still cleaned up whether or not there is an error, so I've added an `if (err < 0)`. Would like your thoughts on this approach (with Go I would've used a `defer` for cleanup, but clearly that is not an option here). There's an alternative approached used in nl.c using 2 goto blocks (https://github.com/lxc/lxc/blob/master/src/lxc/nl.c#L336-L340) but that doesn't use `minus_one_set_errno()` so not sure whether that is using the older style or not.

Thanks